packages/@webex/plugin-meetings/test/unit/spec/media/MediaConnectionAwaiter.ts

```diff
@@ -0,0 +1,326 @@
+
+      let promiseResolved = false;
+      let promiseRejected = false;
+
+      mediaConnectionAwaiter
+        .waitForMediaConnectionConnected()
+        .then(() => {
+          promiseResolved = true;
+        })
+        .catch(() => {
+          promiseRejected = true;
+        });
+
+      assert.equal(promiseResolved, false);
+      assert.equal(promiseRejected, false);
+
+      // check the right listener was registered
+      assert.calledTwice(mockMC.on);
+      assert.equal(mockMC.on.getCall(0).args[0], Event.CONNECTION_STATE_CHANGED);
+      assert.equal(mockMC.on.getCall(1).args[0], Event.ICE_GATHERING_STATE_CHANGED);
+      const listener = mockMC.on.getCall(1).args[1];
+
+      // call the listener and pretend we are now connected
+      mockMC.getIceGatheringState.returns('complete');
+      listener();
+
+      mockMC.getConnectionState.returns(ConnectionState.Connected);
+
+      await clock.tickAsync(ICE_AND_DTLS_CONNECTION_TIMEOUT);
+      await testUtils.flushPromises();
+
+      assert.equal(promiseResolved, true);
+      assert.equal(promiseRejected, false);
+
+      // check that listener was removed
+      assert.calledTwice(mockMC.off);
+
+      assert.calledOnce(clearTimeoutSpy);
+    });
+
+    it(`reject with restart timer once if gathering state is not complete`, async () => {
+      mockMC.getConnectionState.returns(ConnectionState.Connecting);
+      mockMC.getIceGatheringState.returns('new');
+
+      const setTimeoutSpy = sinon.spy(clock, 'setTimeout');
+      const clearTimeoutSpy = sinon.spy(clock, 'clearTimeout');
+
+      let promiseResolved = false;
+      let promiseRejected = false;
+
+      mediaConnectionAwaiter
+        .waitForMediaConnectionConnected()
+        .then(() => {
+          promiseResolved = true;
+        })
+        .catch(() => {
+          promiseRejected = true;
+        });
+
+      assert.equal(promiseResolved, false);
+      assert.equal(promiseRejected, false);
+
+      // check the right listener was registered
+      assert.calledTwice(mockMC.on);
+      assert.equal(mockMC.on.getCall(0).args[0], Event.CONNECTION_STATE_CHANGED);
+      assert.equal(mockMC.on.getCall(1).args[0], Event.ICE_GATHERING_STATE_CHANGED);
+
+      await clock.tickAsync(ICE_AND_DTLS_CONNECTION_TIMEOUT * 3);
+      await testUtils.flushPromises();
+
+      assert.equal(promiseResolved, false);
+      assert.equal(promiseRejected, true);
+
+      // check that listener was removed
+      assert.calledTwice(mockMC.off);
+
+      assert.calledOnce(clearTimeoutSpy);
+      assert.calledTwice(setTimeoutSpy);
+    });
+
+    it(`reject restart timer once if gathering state is not complete`, async () => {
+      mockMC.getConnectionState.returns(ConnectionState.Connecting);
+      mockMC.getIceGatheringState.returns('new');
+
+      const setTimeoutSpy = sinon.spy(clock, 'setTimeout');
+      const clearTimeoutSpy = sinon.spy(clock, 'clearTimeout');
+
+      let promiseResolved = false;
```

Comment:
just a suggestion, but there is a lot of repeated code in all these tests (lines 290-311 in this one or even more), maybe it could be extracted into a common function?

Reply:
Tried to split it a little bit into subtests with common beforeEach/afterEach, but it was harder to read and understand the tests. If you don't mind I would suggest keeping it as it is.

Reply:
ok, fine